### PR TITLE
support GPU waiting when no stream memory operations

### PR DESCRIPTION
**Pull request by benson31**
An implementation of GPU waiting semantics that does not rely on stream memory operations. There are concerns about this implementation, but it seems to work on pascal.

Combined with #2, I now see consistent and better performance.

### CODE REVIEW

File: src/cuda.cpp [hunk 1]
```diff
@@ -108,46 +109,78 @@ bool stream_memory_operations_supported() {
...
   // Can't use std::atomic because we need the actual address.
   __atomic_store_n(wait_sync, 0, __ATOMIC_SEQ_CST);
-  AL_CHECK_CUDA_DRV(cuMemHostGetDevicePointer(
-                      &wait_sync_dev_ptr, wait_sync, 0));
+
+  if (stream_memory_operations_supported())
+    AL_CHECK_CUDA_DRV(cuMemHostGetDevicePointer(
+                        &wait_sync_dev_ptr, wait_sync, 0));
+  else
+    AL_CHECK_CUDA(cudaHostGetDevicePointer(
```
Comment (naoyam) [inline src/cuda.cpp]: Why does it use cudaHostGetDevicePointer here and cuMemHostGetDevicePointer when stream memory operations are supported?

File: src/cuda.hpp [hunk 2]
```diff
@@ -172,6 +178,7 @@ class GPUWait {
   void signal();
  private:
   int32_t* wait_sync __attribute__((aligned(64)));
```
Comment (naoyam) [inline src/cuda.hpp]: This is not part of your change, but how does the aligned attribute work here? I assume it is to make the address of wait_sync be aligned at 64-byte boundaries, does it make any difference? Also, how this is implemented when the variable is actually a field of a class? Does a compiler fill proper padding when necessary?

File: src/cuda.cpp (hunk 1 above)
Comment (ndryden) [reply src/cuda.cpp]: The runtime and driver APIs actually represent device pointers slightly differently. The runtime API just treats them as pointers, but the driver API uses the `CUdeviceptr` type.

File: src/cuda.hpp (hunk 2 above)
Comment (ndryden) [reply src/cuda.hpp]: I haven't benchmarked it in detail, but it's meant to avoid false sharing in cache lines. The compiler will pad the structure/class as needed to enforce the alignment (this can change the size of the struct/class).

File: src/cuda.cpp (hunk 1 above)
Comment (naoyam) [reply src/cuda.cpp]: OK, thanks.

**Comment by benson31** [comment]
I agree about tests. Right now I'm treating LBANN as the test, but I'd like to spend a week or two on infrastructure once things calm down. I am going to add tests for alltoall, gather, and scatter, but I'd also like to unit test these components, GPUWait and FastEvent at a minimum.

File: src/cuda.cpp [hunk 3]
```diff
@@ -108,46 +109,78 @@ bool stream_memory_operations_supported() {
...
 void GPUWait::wait(cudaStream_t stream) {
-  AL_CHECK_CUDA_DRV(cuStreamWaitValue32(
-                      stream, wait_sync_dev_ptr, 1, CU_STREAM_WAIT_VALUE_EQ));
+  if (stream_memory_operations_supported())
+    launch_wait_kernel(stream, 1, wait_sync_dev_ptr);
+  else
+    launch_wait_kernel(stream, 1, wait_sync_dev_ptr_no_stream_mem_ops);
 }
 
 void GPUWait::signal() {
```
Comment (naoyam) [inline src/cuda.cpp]: Have you tested write-combined host memory? Since the memory is write only at host and read only at device, write-combined pinned memory may be more efficient. This also applies regardless of the availability of the stream operations.

Comment (ndryden) [reply src/cuda.cpp]: This would be good to test. I've considered it but haven't gotten to it-- I'm somewhat concerned that write-combined memory will have better bandwidth but worse latency.

File: src/cuda.hpp [hunk 4]
```diff
@@ -172,6 +178,7 @@ class GPUWait {
...
+  int32_t* wait_sync_dev_ptr_no_stream_mem_ops __attribute__((aligned(64)));
```
Comment (ndryden) [inline src/cuda.hpp]: Something to consider later would be to see if we can do this without additional memory overhead.

File: src/cuda.cpp (hunk 3 above)
Comment (naoyam) [reply src/cuda.cpp]: Which latency are you referring to? My understanding is that it should just make the caching of the memory write-through instead of write-back. Read latency at the device should be smaller. Writes on host would be more costly, but it's just one time operation.

Comment (ndryden) [reply src/cuda.cpp]: I had assumed that write-combined memory temporarily stored the writes to e.g. a cache line in a buffer before releasing them later in a burst mode, which should increase latency. But I could easily be wrong, I haven't looked in detail at it.

Comment (naoyam) [reply src/cuda.cpp]: I didn't think about that. That may be right.

File: src/cuda.hpp (hunk 4 above)
Comment (benson31) [reply src/cuda.hpp]: With the unified memory stuff, I don't think you actually need to get the device pointer. It's probably the same if I'm reading [this doc](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY.html#group__CUDART__MEMORY_1gc00502b44e5f1bdc0b424487ebb08db0) correctly. This is all also possible with just the driver API, in which case CUdeviceptr is probably what we'd need, but I'm less comfortable with that API.
